File: packages/contracts-ts/src/canonical.ts
/**
 * Canonical JSON (`cognitiveos.canonical-json/0.1`) and domain-separated
 * digest / signature preimages — TypeScript twin of
 * `crates/cognitive-contracts/src/canonical.rs`.
 *
 * Implements `docs/standards/canonical-encoding-and-digest.md` sections 2-9
 * and 12 for the encoding layer: UTF-8 without BOM, I-JSON, RFC 8785 JCS
 * output, duplicate-member rejection, unsafe-integer rejection, and the
 * `CognitiveOS-Digest-V1` / `CognitiveOS-Signature-V1` preimages.
 *
 * RFC 8785 note: JCS string escaping and number serialization are defined in
 * terms of ECMAScript `JSON.stringify` / `ToString(Number)`, so the native
 * primitives are used directly; ordering is by UTF-16 code units, which is
 * plain JavaScript string comparison.
 */

import { createHash } from "node:crypto";

/** I-JSON safe integer bound (RFC 7493): 2^53 - 1. */
export const MAX_SAFE_INTEGER_LITERAL = "9007199254740991";

/** Exact ASCII prefix of every digest preimage (standard section 9). */
export const DIGEST_PREIMAGE_PREFIX = "CognitiveOS-Digest-V1\n";

/** Exact ASCII prefix of every signature input (standard section 12). */
export const SIGNATURE_PREIMAGE_PREFIX = "CognitiveOS-Signature-V1\n";

/**
 * Rejection categories shared with the Rust implementation and the golden
 * fixtures under `tests/golden/`. String forms are fixture contract.
 */
export type CanonicalErrorCategory =
  | "invalid-utf8"
  | "bom"
  | "duplicate-member-name"
  | "unsafe-integer"
  | "invalid-json"
  | "invalid-domain"
  | "invalid-algorithm"
  | "canonicalization-failed";

export class CanonicalError extends Error {
  readonly category: CanonicalErrorCategory;

  constructor(category: CanonicalErrorCategory, detail: string) {
    super(detail ? `${category}: ${detail}` : category);
    this.name = "CanonicalError";
    this.category = category;
  }
}

/** Strict JSON value. Objects preserve document member order. */
export type StrictValue =
  | null
  | boolean
  | number
  | string
  | StrictValue[]
  | { readonly kind: "object"; readonly members: ReadonlyArray<readonly [string, StrictValue]> };

function isObjectValue(
  value: StrictValue,
): value is { readonly kind: "object"; readonly members: ReadonlyArray<readonly [string, StrictValue]> } {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Strict recursive-descent JSON parser (standard section 3): rejects BOM,
 * duplicate member names, unsafe integer literals, unpaired surrogate
 * escapes and every JSON grammar violation. `JSON.parse` cannot be used
 * because it silently keeps the last duplicate member and widens integers.
 */
class StrictParser {
  private pos = 0;

  constructor(private readonly text: string) {}

  parse(): StrictValue {
    this.skipWs();
    const value = this.parseValue();
    this.skipWs();
    if (this.pos !== this.text.length) {
      throw new CanonicalError("invalid-json", `trailing characters at offset ${this.pos}`);
    }
    return value;
  }

  private fail(reason: string): never {
    throw new CanonicalError("invalid-json", `${reason} at offset ${this.pos}`);
  }

  private skipWs(): void {
    while (this.pos < this.text.length) {
      const c = this.text[this.pos];
      if (c === " " || c === "\t" || c === "\n" || c === "\r") {
        this.pos += 1;
      } else {
        break;
      }
    }
  }

  private parseValue(): StrictValue {
    const c = this.text[this.pos];
    if (c === undefined) {
      this.fail("unexpected end of input");
    }
    switch (c) {
      case "{":
        return this.parseObject();
      case "[":
        return this.parseArray();
      case '"':
        return this.parseString();
      case "t":
        this.expectLiteral("true");
        return true;
      case "f":
        this.expectLiteral("false");
        return false;
      case "n":
        this.expectLiteral("null");
        return null;
      default:
        if (c === "-" || (c >= "0" && c <= "9")) {
          return this.parseNumber();
        }
        return this.fail(`unexpected character ${JSON.stringify(c)}`);
    }
  }

  private expectLiteral(literal: string): void {
    if (this.text.startsWith(literal, this.pos)) {
      this.pos += literal.length;
    } else {
      this.fail(`invalid literal, expected ${literal}`);
    }
  }

  private parseObject(): StrictValue {
    this.pos += 1; // consume '{'
    const members: Array<readonly [string, StrictValue]> = [];
    const seen = new Set<string>();
    this.skipWs();
    if (this.text[this.pos] === "}") {
      this.pos += 1;
      return { kind: "object", members };
    }
    for (;;) {
      this.skipWs();
      if (this.text[this.pos] !== '"') {
        this.fail("expected member name string");
      }
      const name = this.parseString();
      if (seen.has(name)) {
        throw new CanonicalError("duplicate-member-name", name);
      }
      seen.add(name);
      this.skipWs();
      if (this.text[this.pos] !== ":") {
        this.fail("expected ':' after member name");
      }
      this.pos += 1;
      this.skipWs();
      members.push([name, this.parseValue()] as const);
      this.skipWs();
      const next = this.text[this.pos];
      if (next === ",") {
        this.pos += 1;
        continue;
      }
      if (next === "}") {
        this.pos += 1;
        return { kind: "object", members };
      }
      this.fail("expected ',' or '}' in object");
    }
  }

  private parseArray(): StrictValue {
    this.pos += 1; // consume '['
    const items: StrictValue[] = [];
    this.skipWs();
    if (this.text[this.pos] === "]") {
      this.pos += 1;
      return items;
    }
    for (;;) {
      this.skipWs();
      items.push(this.parseValue());
      this.skipWs();
      const next = this.text[this.pos];
      if (next === ",") {
        this.pos += 1;
        continue;
      }
      if (next === "]") {
        this.pos += 1;
        return items;
      }
      this.fail("expected ',' or ']' in array");
    }
  }

  private parseString(): string {
    this.pos += 1; // consume opening quote
    let out = "";
    for (;;) {
      const c = this.text[this.pos];
      if (c === undefined) {
        this.fail("unterminated string");
      }
      if (c === '"') {
        this.pos += 1;
        return out;
      }
      if (c === "\\") {
        out += this.parseEscape();
        continue;
      }
      const code = c.codePointAt(0) ?? 0;
      if (code < 0x20) {
        this.fail("unescaped control character in string");
      }
      // Unpaired surrogates cannot survive UTF-8 decoding with fatal=true,
      // but guard against direct string inputs.
      if (code >= 0xd800 && code <= 0xdfff) {
        const pair = this.text.codePointAt(this.pos) ?? 0;
        if (pair >= 0xd800 && pair <= 0xdfff) {
          this.fail("unpaired surrogate in string");
        }
        const composed = String.fromCodePoint(pair);
        out += composed;
        this.pos += composed.length;
        continue;
      }
      out += c;
      this.pos += 1;
    }
  }

  private parseEscape(): string {
    this.pos += 1; // consume backslash
    const c = this.text[this.pos];
    this.pos += 1;
    switch (c) {
      case '"':
        return '"';
      case "\\":
        return "\\";
      case "/":
        return "/";
      case "b":
        return "\b";
      case "f":
        return "\f";
      case "n":
        return "\n";
      case "r":
        return "\r";
      case "t":
        return "\t";
      case "u": {
        const unit = this.parseHex4();
        if (unit >= 0xd800 && unit <= 0xdbff) {
          // Expect a low-surrogate escape to complete the pair.
          if (this.text[this.pos] === "\\" && this.text[this.pos + 1] === "u") {
            this.pos += 2;
            const low = this.parseHex4();
            if (low >= 0xdc00 && low <= 0xdfff) {
              return String.fromCharCode(unit, low);
            }
          }
          this.fail("unpaired surrogate escape");
        }
        if (unit >= 0xdc00 && unit <= 0xdfff) {
          this.fail("unpaired surrogate escape");
        }
        return String.fromCharCode(unit);
      }
      default:
        return this.fail("invalid escape sequence");
    }
  }

  private parseHex4(): number {
    const hex = this.text.slice(this.pos, this.pos + 4);
    if (!/^[0-9a-fA-F]{4}$/.test(hex)) {
      this.fail("invalid \\u escape");
    }
    this.pos += 4;
    return Number.parseInt(hex, 16);
  }

  private parseNumber(): number {
    const start = this.pos;
    const rest = this.text.slice(start);
    const match = /^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/.exec(rest);
    if (!match || match[0].length === 0) {
      this.fail("invalid number literal");
    }
    const literal = match[0];
    this.pos += literal.length;
    const isIntegerForm = !/[.eE]/.test(literal);
    if (isIntegerForm) {
      const digits = literal.startsWith("-") ? literal.slice(1) : literal;
      const tooBig =
        digits.length > MAX_SAFE_INTEGER_LITERAL.length ||
        (digits.length === MAX_SAFE_INTEGER_LITERAL.length && digits > MAX_SAFE_INTEGER_LITERAL);
      if (tooBig) {
        throw new CanonicalError("unsafe-integer", literal);
      }
    }
    const value = Number(literal);
    if (!Number.isFinite(value)) {
      throw new CanonicalError("unsafe-integer", literal);
    }
    return value;
  }
}

/** Strictly parse a JSON text (standard sections 3-5). */
export function parseStrict(input: string): StrictValue {
  if (input.startsWith("\uFEFF")) {
    throw new CanonicalError("bom", "input begins with a byte order mark");
  }
  return new StrictParser(input).parse();
}

/** Strictly decode UTF-8 bytes, then parse. Invalid UTF-8 and BOM are rejected. */
export function parseStrictBytes(input: Uint8Array): StrictValue {
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(input);
  } catch {
    throw new CanonicalError("invalid-utf8", "input is not well-formed UTF-8");
  }
  return parseStrict(text);
}

function serializeCanonical(value: StrictValue): string {
  if (value === null) {
    return "null";
  }
  switch (typeof value) {
    case "boolean":
      return value ? "true" : "false";
    case "number": {
      if (!Number.isFinite(value)) {
        throw new CanonicalError("canonicalization-failed", "non-finite number");
      }
      // RFC 8785: ECMAScript ToString(Number); ToString(-0) is "0".
      return Object.is(value, -0) ? "0" : String(value);
    }
    case "string":
      // RFC 8785 string serialization is ECMAScript JSON.stringify quoting.
      return JSON.stringify(value);
    default:
      break;
  }
  if (Array.isArray(value)) {
    return `[${value.map(serializeCanonical).join(",")}]`;
  }
  if (isObjectValue(value)) {
    // JCS: sort member names by UTF-16 code units (plain JS comparison).
    const sorted = [...value.members].sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
    const body = sorted
      .map(([name, member]) => `${JSON.stringify(name)}:${serializeCanonical(member)}`)
      .join(",");
    return `{${body}}`;
  }
  throw new CanonicalError("canonicalization-failed", "unsupported value");
}

/** Produce RFC 8785 canonical UTF-8 bytes for a strict-parsed value. */
export function canonicalBytes(value: StrictValue): Uint8Array {
  return new TextEncoder().encode(serializeCanonical(value));
}

/** Strict-parse a JSON text and return its RFC 8785 canonical bytes. */
export function canonicalize(input: string): Uint8Array {
  return canonicalBytes(parseStrict(input));
}

const DOMAIN_GRAMMAR = /^[a-z0-9][a-z0-9._/-]{0,127}$/;
const FORBIDDEN_DOMAINS = new Set(["generic", "object", "payload"]);

/** Validate a digest/signature domain label (standard section 9). */
export function validateDomain(domain: string): void {
  if (!DOMAIN_GRAMMAR.test(domain) || FORBIDDEN_DOMAINS.has(domain)) {
    throw new CanonicalError("invalid-domain", domain);
  }
}

function concatBytes(parts: ReadonlyArray<Uint8Array>): Uint8Array {
  const total = parts.reduce((sum, part) => sum + part.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

const ascii = (text: string): Uint8Array => new TextEncoder().encode(text);

/** Exact digest preimage: `"CognitiveOS-Digest-V1\n" || domain || 0x00 || C`. */
export function digestPreimage(canonical: Uint8Array, domain: string): Uint8Array {
  validateDomain(domain);
  return concatBytes([ascii(DIGEST_PREIMAGE_PREFIX), ascii(domain), Uint8Array.of(0), canonical]);
}

/** Domain-separated digest: `"sha256:" || lowercase_hex(SHA-256(preimage))`. */
export function digest(canonical: Uint8Array, domain: string): string {
  const hash = createHash("sha256").update(digestPreimage(canonical, domain)).digest("hex");
  return `sha256:${hash}`;
}

/**
 * Exact signature input (standard section 12):
 * `"CognitiveOS-Signature-V1\n" || domain || 0x00 || algorithm || 0x00 || C`.
 */
export function signatureInput(canonical: Uint8Array, domain: string, algorithm: string): Uint8Array {
  validateDomain(domain);
  if (algorithm.length === 0 || !/^[\x00-\x7F]+$/.test(algorithm)) {
    throw new CanonicalError("invalid-algorithm", algorithm);
  }
  return concatBytes([
    ascii(SIGNATURE_PREIMAGE_PREFIX),
    ascii(domain),
    Uint8Array.of(0),
    ascii(algorithm),
    Uint8Array.of(0),
    canonical,
  ]);
}

/** Convenience: strict parse + canonicalize + domain-separated digest. */
export function digestJson(input: string, domain: string): string {
  return digest(canonicalize(input), domain);
}

File: crates/cognitive-contracts/src/canonical.rs
//! Canonical JSON (`cognitiveos.canonical-json/0.1`) and domain-separated
//! digest / signature preimages.
//!
//! Implements `docs/standards/canonical-encoding-and-digest.md` sections 2-9
//! and 12 for the encoding layer: UTF-8 without BOM, I-JSON, RFC 8785 JCS
//! output, duplicate-member rejection, unsafe-integer rejection, and the
//! `CognitiveOS-Digest-V1` / `CognitiveOS-Signature-V1` preimages.
//!
//! Schema-bound rules (timestamp field contracts, digest projections, set
//! manifests) are enforced one level up, next to schema validation; they are
//! not part of this encoding-layer module.

use serde::Deserialize;
use serde::de::{self, Deserializer, MapAccess, SeqAccess, Visitor};
use serde::ser::{Serialize, SerializeMap, SerializeSeq, Serializer};
use sha2::{Digest, Sha256};
use std::fmt;

/// I-JSON safe integer bound (RFC 7493): 2^53 - 1.
pub const MAX_SAFE_INTEGER: i64 = 9_007_199_254_740_991;

/// Exact ASCII prefix of every digest preimage (standard section 9).
pub const DIGEST_PREIMAGE_PREFIX: &[u8] = b"CognitiveOS-Digest-V1\n";

/// Exact ASCII prefix of every signature input (standard section 12).
pub const SIGNATURE_PREIMAGE_PREFIX: &[u8] = b"CognitiveOS-Signature-V1\n";

/// Rejection categories shared with the TypeScript implementation and the
/// golden fixtures under `tests/golden/`. String forms are fixture contract.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum CanonicalError {
    /// Input bytes are not well-formed shortest-form UTF-8.
    #[error("invalid-utf8")]
    InvalidUtf8,
    /// Input begins with a U+FEFF byte order mark.
    #[error("bom")]
    Bom,
    /// An object contains the same member name twice.
    #[error("duplicate-member-name: {0}")]
    DuplicateMemberName(String),
    /// An exact integer literal outside the I-JSON safe range used a JSON number.
    #[error("unsafe-integer: {0}")]
    UnsafeInteger(String),
    /// Any other JSON grammar violation (NaN/Infinity literals, lone
    /// surrogates in escapes, trailing content, truncation, ...).
    #[error("invalid-json: {0}")]
    InvalidJson(String),
    /// Digest/signature domain label violates the registered grammar.
    #[error("invalid-domain: {0}")]
    InvalidDomain(String),
    /// Signature algorithm identifier is empty or not ASCII.
    #[error("invalid-algorithm: {0}")]
    InvalidAlgorithm(String),
    /// RFC 8785 serialization failed (non-finite number reached the encoder).
    #[error("canonicalization-failed: {0}")]
    CanonicalizationFailed(String),
}

impl CanonicalError {
    /// Stable machine category used by golden fixtures (`expected_rejection`).
    pub fn category(&self) -> &'static str {
        match self {
            CanonicalError::InvalidUtf8 => "invalid-utf8",
            CanonicalError::Bom => "bom",
            CanonicalError::DuplicateMemberName(_) => "duplicate-member-name",
            CanonicalError::UnsafeInteger(_) => "unsafe-integer",
            CanonicalError::InvalidJson(_) => "invalid-json",
            CanonicalError::InvalidDomain(_) => "invalid-domain",
            CanonicalError::InvalidAlgorithm(_) => "invalid-algorithm",
            CanonicalError::CanonicalizationFailed(_) => "canonicalization-failed",
        }
    }
}

// Marker prefixes smuggled through serde's custom error channel from the
// strict visitor, then mapped back to typed variants in `parse_strict`.
const DUP_MARKER: &str = "__COS_DUPLICATE__:";
const UNSAFE_MARKER: &str = "__COS_UNSAFE_INT__:";

/// Strict JSON value: preserves document member order, rejects duplicate
/// member names and unsafe integers at parse time.
#[derive(Debug, Clone, PartialEq)]
pub enum StrictValue {
    Null,
    Bool(bool),
    /// Exact integer within the I-JSON safe range.
    Int(i64),
    /// Any other finite binary64 number.
    Float(f64),
    String(String),
    Array(Vec<StrictValue>),
    /// Members in document order; JCS ordering is applied at serialization.
    Object(Vec<(String, StrictValue)>),
}

impl<'de> Deserialize<'de> for StrictValue {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct StrictVisitor;

        impl<'de> Visitor<'de> for StrictVisitor {
            type Value = StrictValue;

            fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
                f.write_str("any strict I-JSON value")
            }

            fn visit_bool<E>(self, v: bool) -> Result<StrictValue, E> {
                Ok(StrictValue::Bool(v))
            }

            fn visit_i64<E: de::Error>(self, v: i64) -> Result<StrictValue, E> {
                if v.unsigned_abs() > MAX_SAFE_INTEGER as u64 {
                    return Err(de::Error::custom(format!("{UNSAFE_MARKER}{v}")));
                }
                Ok(StrictValue::Int(v))
            }

            fn visit_u64<E: de::Error>(self, v: u64) -> Result<StrictValue, E> {
                if v > MAX_SAFE_INTEGER as u64 {
                    return Err(de::Error::custom(format!("{UNSAFE_MARKER}{v}")));
                }
                Ok(StrictValue::Int(v as i64))
            }

            fn visit_f64<E>(self, v: f64) -> Result<StrictValue, E> {
                Ok(StrictValue::Float(v))
            }

            fn visit_str<E>(self, v: &str) -> Result<StrictValue, E> {
                Ok(StrictValue::String(v.to_owned()))
            }

            fn visit_string<E>(self, v: String) -> Result<StrictValue, E> {
                Ok(StrictValue::String(v))
            }

            fn visit_unit<E>(self) -> Result<StrictValue, E> {
                Ok(StrictValue::Null)
            }

            fn visit_seq<A: SeqAccess<'de>>(self, mut seq: A) -> Result<StrictValue, A::Error> {
                let mut items = Vec::new();
                while let Some(item) = seq.next_element::<StrictValue>()? {
                    items.push(item);
                }
                Ok(StrictValue::Array(items))
            }

            fn visit_map<A: MapAccess<'de>>(self, mut map: A) -> Result<StrictValue, A::Error> {
                let mut members: Vec<(String, StrictValue)> = Vec::new();
                while let Some(key) = map.next_key::<String>()? {
                    if members.iter().any(|(existing, _)| existing == &key) {
                        return Err(de::Error::custom(format!("{DUP_MARKER}{key}")));
                    }
                    let value = map.next_value::<StrictValue>()?;
                    members.push((key, value));
                }
                Ok(StrictValue::Object(members))
            }
        }

        deserializer.deserialize_any(StrictVisitor)
    }
}

impl Serialize for StrictValue {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        match self {
            StrictValue::Null => serializer.serialize_unit(),
            StrictValue::Bool(b) => serializer.serialize_bool(*b),
            StrictValue::Int(i) => serializer.serialize_i64(*i),
            StrictValue::Float(f) => serializer.serialize_f64(*f),
            StrictValue::String(s) => serializer.serialize_str(s),
            StrictValue::Array(items) => {
                let mut seq = serializer.serialize_seq(Some(items.len()))?;
                for item in items {
                    seq.serialize_element(item)?;
                }
                seq.end()
            }
            StrictValue::Object(members) => {
                let mut map = serializer.serialize_map(Some(members.len()))?;
                for (key, value) in members {
                    map.serialize_entry(key, value)?;
                }
                map.end()
            }
        }
    }
}

/// Strictly parse UTF-8 bytes into a [`StrictValue`] (standard sections 3-5).
pub fn parse_strict_bytes(input: &[u8]) -> Result<StrictValue, CanonicalError> {
    let text = std::str::from_utf8(input).map_err(|_| CanonicalError::InvalidUtf8)?;
    parse_strict(text)
}

/// Reject exact-integer literals outside the I-JSON safe range before the
/// main parse. serde_json silently widens integer literals beyond u64/i64 to
/// binary64, which would violate standard section 5, so this lexical scan
/// runs first. String bodies (including escapes) are skipped; in valid JSON
/// every digit run outside a string belongs to a number literal.
fn reject_unsafe_integer_literals(input: &str) -> Result<(), CanonicalError> {
    const SAFE_DIGITS: usize = 16; // "9007199254740991".len()
    let bytes = input.as_bytes();
    let mut i = 0usize;
    let mut in_string = false;
    while i < bytes.len() {
        let b = bytes[i];
        if in_string {
            match b {
                b'\\' => i += 2,
                b'"' => {
                    in_string = false;
                    i += 1;
                }
                _ => i += 1,
            }
            continue;
        }
        match b {
            b'"' => {
                in_string = true;
                i += 1;
            }
            b'-' | b'0'..=b'9' => {
                let start = i;
                while i < bytes.len()
                    && matches!(bytes[i], b'-' | b'+' | b'.' | b'e' | b'E' | b'0'..=b'9')
                {
                    i += 1;
                }
                let literal = &input[start..i];
                let is_integer_form = !literal.bytes().any(|c| matches!(c, b'.' | b'e' | b'E'));
                if is_integer_form {
                    let digits = literal.strip_prefix('-').unwrap_or(literal);
                    let too_big = digits.len() > SAFE_DIGITS
                        || (digits.len() == SAFE_DIGITS && digits > "9007199254740991");
                    if too_big {
                        return Err(CanonicalError::UnsafeInteger(literal.to_owned()));
                    }
                }
            }
            _ => i += 1,
        }
    }
    Ok(())
}

/// Strictly parse a JSON text into a [`StrictValue`].
pub fn parse_strict(input: &str) -> Result<StrictValue, CanonicalError> {
    if input.starts_with('\u{FEFF}') {
        return Err(CanonicalError::Bom);
    }
    reject_unsafe_integer_literals(input)?;
    serde_json::from_str::<StrictValue>(input).map_err(|err| {
        let msg = err.to_string();
        if let Some(pos) = msg.find(DUP_MARKER) {
            let rest = &msg[pos + DUP_MARKER.len()..];
            let name = rest.split(" at line").next().unwrap_or(rest);
            return CanonicalError::DuplicateMemberName(name.to_owned());
        }
        if let Some(pos) = msg.find(UNSAFE_MARKER) {
            let rest = &msg[pos + UNSAFE_MARKER.len()..];
            let lit = rest.split(" at line").next().unwrap_or(rest);
            return CanonicalError::UnsafeInteger(lit.to_owned());
        }
        CanonicalError::InvalidJson(msg)
    })
}

/// Produce RFC 8785 canonical bytes for an already strict-parsed value.
pub fn canonical_bytes(value: &StrictValue) -> Result<Vec<u8>, CanonicalError> {
    serde_json_canonicalizer::to_vec(value)
        .map_err(|err| CanonicalError::CanonicalizationFailed(err.to_string()))
}

/// Strict-parse a JSON text and return its RFC 8785 canonical bytes.
pub fn canonicalize(input: &str) -> Result<Vec<u8>, CanonicalError> {
    canonical_bytes(&parse_strict(input)?)
}

/// Canonicalize an in-memory `serde_json::Value`. Duplicate member names are
/// unrepresentable in `Value`; unsafe integers are still rejected so a value
/// is never silently rounded before a digest (standard section 5).
pub fn canonical_bytes_of_value(value: &serde_json::Value) -> Result<Vec<u8>, CanonicalError> {
    reject_unsafe_value_numbers(value)?;
    serde_json_canonicalizer::to_vec(value)
        .map_err(|err| CanonicalError::CanonicalizationFailed(err.to_string()))
}

fn reject_unsafe_value_numbers(value: &serde_json::Value) -> Result<(), CanonicalError> {
    match value {
        serde_json::Value::Number(number) => {
            let unsafe_int = number
                .as_i64()
                .map(|i| i.unsigned_abs() > MAX_SAFE_INTEGER as u64)
                .or_else(|| number.as_u64().map(|u| u > MAX_SAFE_INTEGER as u64))
                .unwrap_or(false);
            if unsafe_int {
                return Err(CanonicalError::UnsafeInteger(number.to_string()));
            }
            Ok(())
        }
        serde_json::Value::Array(items) => items.iter().try_for_each(reject_unsafe_value_numbers),
        serde_json::Value::Object(members) => {
            members.values().try_for_each(reject_unsafe_value_numbers)
        }
        _ => Ok(()),
    }
}

/// Validate a digest/signature domain label (standard section 9).
pub fn validate_domain(domain: &str) -> Result<(), CanonicalError> {
    const FORBIDDEN: [&str; 3] = ["generic", "object", "payload"];
    let bytes = domain.as_bytes();
    let valid_head = |b: u8| b.is_ascii_lowercase() || b.is_ascii_digit();
    let valid_tail = |b: u8| valid_head(b) || b == b'.' || b == b'_' || b == b'/' || b == b'-';
    let grammar_ok = match bytes.split_first() {
        Some((head, tail)) => {
            bytes.len() <= 128 && valid_head(*head) && tail.iter().all(|b| valid_tail(*b))
        }
        None => false,
    };
    if !grammar_ok || FORBIDDEN.contains(&domain) {
        return Err(CanonicalError::InvalidDomain(domain.to_owned()));
    }
    Ok(())
}

/// Exact digest preimage: `"CognitiveOS-Digest-V1\n" || domain || 0x00 || C`.
pub fn digest_preimage(canonical: &[u8], domain: &str) -> Result<Vec<u8>, CanonicalError> {
    validate_domain(domain)?;
    let mut preimage =
        Vec::with_capacity(DIGEST_PREIMAGE_PREFIX.len() + domain.len() + 1 + canonical.len());
    preimage.extend_from_slice(DIGEST_PREIMAGE_PREFIX);
    preimage.extend_from_slice(domain.as_bytes());
    preimage.push(0x00);
    preimage.extend_from_slice(canonical);
    Ok(preimage)
}

/// Domain-separated digest: `"sha256:" || lowercase_hex(SHA-256(preimage))`.
pub fn digest(canonical: &[u8], domain: &str) -> Result<String, CanonicalError> {
    let preimage = digest_preimage(canonical, domain)?;
    Ok(format!(
        "sha256:{}",
        lowercase_hex(&Sha256::digest(&preimage))
    ))
}

/// Exact signature input (standard section 12):
/// `"CognitiveOS-Signature-V1\n" || domain || 0x00 || algorithm || 0x00 || C`.
pub fn signature_input(
    canonical: &[u8],
    domain: &str,
    algorithm: &str,
) -> Result<Vec<u8>, CanonicalError> {
    validate_domain(domain)?;
    if algorithm.is_empty() || !algorithm.is_ascii() {
        return Err(CanonicalError::InvalidAlgorithm(algorithm.to_owned()));
    }
    let mut input = Vec::with_capacity(
        SIGNATURE_PREIMAGE_PREFIX.len() + domain.len() + algorithm.len() + 2 + canonical.len(),
    );
    input.extend_from_slice(SIGNATURE_PREIMAGE_PREFIX);
    input.extend_from_slice(domain.as_bytes());
    input.push(0x00);
    input.extend_from_slice(algorithm.as_bytes());
    input.push(0x00);
    input.extend_from_slice(canonical);
    Ok(input)
}

/// Convenience: strict parse + canonicalize + domain-separated digest.
pub fn digest_json(input: &str, domain: &str) -> Result<String, CanonicalError> {
    digest(&canonicalize(input)?, domain)
}

fn lowercase_hex(bytes: &[u8]) -> String {
    let mut out = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        use fmt::Write;
        // Writing to a String cannot fail; ignore the Infallible result.
        let _ = write!(out, "{b:02x}");
    }
    out
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::panic)]
mod tests {
    use super::*;

    #[test]
    fn rejects_duplicate_member_names() {
        let err = parse_strict(r#"{"a":1,"a":2}"#).unwrap_err();
        assert_eq!(err.category(), "duplicate-member-name");
    }

    #[test]
    fn rejects_unsafe_integers() {
        let err = parse_strict("9007199254740993").unwrap_err();
        assert_eq!(err.category(), "unsafe-integer");
        assert!(parse_strict("9007199254740991").is_ok());
        assert!(parse_strict("-9007199254740991").is_ok());
    }

    #[test]
    fn rejects_bom_and_invalid_utf8() {
        assert_eq!(parse_strict("\u{FEFF}{}").unwrap_err().category(), "bom");
        assert_eq!(
            parse_strict_bytes(&[0xFF, 0xFE]).unwrap_err().category(),
            "invalid-utf8"
        );
    }

    #[test]
    fn canonical_key_ordering_and_unicode() {
        let bytes = canonicalize(r#"{"b":2,"a":1,"\u20ac":"euro"}"#).unwrap();
        assert_eq!(
            String::from_utf8(bytes).unwrap(),
            "{\"a\":1,\"b\":2,\"\u{20ac}\":\"euro\"}"
        );
    }

    #[test]
    fn digest_uses_domain_separation() {
        let canonical = canonicalize("{}").unwrap();
        let d1 = digest(&canonical, "conformance-fixture/0.1").unwrap();
        let d2 = digest(&canonical, "schema-bundle/0.1").unwrap();
        assert_ne!(d1, d2);
        assert!(d1.starts_with("sha256:") && d1.len() == 7 + 64);
    }

    #[test]
    fn forbidden_domains_rejected() {
        let canonical = canonicalize("{}").unwrap();
        for domain in ["", "generic", "object", "payload", "UPPER", "-lead"] {
            assert_eq!(
                digest(&canonical, domain).unwrap_err().category(),
                "invalid-domain",
                "domain {domain:?} must be rejected"
            );
        }
    }
}
